src/forder.c

```diff
@@ -307,20 +302,22 @@ static void range_str(const SEXP *x, int n, uint64_t *out_min, uint64_t *out_max
       na_count++;
       continue;
     }
-    if (TRUELENGTH(s)<0) continue;  // seen this group before
-    #pragma omp critical
-    if (TRUELENGTH(s)>=0) {  // another thread may have set it while I was waiting, so check it again
-      if (TRUELENGTH(s)>0)   // save any of R's own usage of tl (assumed positive, so we can both count and save in one scan), to restore
-        savetl(s);           // afterwards. From R 2.14.0, tl is initialized to 0, prior to that it was random so this step saved too much.
-      // now save unique SEXP in ustr so i) we can loop through them afterwards and reset TRUELENGTH to 0 and ii) sort uniques when sorting too
+    // Why is it acceptable to call dhash_lookup when marks can be shared between threads?
```

Comment:
So far the only way I know to make this provably safe is to throw away old hash table buffers (to be garbage collected by R).

Fast, thread-safe, dynamically growing, memory efficient hash table: choose two out of four, maybe three if you're lucky.